### PR TITLE
[Hybrid] Warmup Mamba2 SSD kernel

### PR DESCRIPTION
**Pull request by tdoublep**
## Summary

Triton's auto-tuner for the Mamba2 SSD kernels currently runs lazily on the first inference request, causing a large latency spike. This PR adds a `_warmup_ssd_kernels()` method to `MambaMixer2` that triggers auto-tuning during vLLM's profile phase (before SSM cache allocation), shifting the cost into server startup.

- Runs a minimal `mamba_chunk_scan_combined_varlen` forward pass with dummy tensors during the V1 profile run
- Covers both `HAS_INITSTATES` constexpr code paths (with and without `initial_states`)
- Uses correct dtypes (including `ssm_state_dtype`) to match Triton's cache keys
- Each layer warms up once, and subsequent layers hit the Triton cache

## Benchmark Results

Measured on a single H100 80GB with `nvidia/NVIDIA-Nemotron-3-Nano-4B-BF16` (Mamba2 hybrid model), `max_model_len=512`, cold Triton cache (`TRITON_PRINT_AUTOTUNING=1` confirmed autotuning location):

| Metric | Baseline (main) | With Warmup | Change |
|--------|-----------------|-------------|--------|
| Model load time | 30.0s | 76.9s | +46.9s (autotuning shifted here) |
| First request latency | 31.343s | 2.890s | **-28.5s (91% reduction)** |
| Subsequent request avg | 0.083s | 0.083s | No change |
| First / subsequent ratio | 378x | 35x | **10.8x improvement** |

With `TRITON_PRINT_AUTOTUNING=1`, 58 kernel autotuning events occurred **after** model load on main (during the first request). With the warmup branch, **zero** autotuning events occurred after model load -- all SSD kernel autotuning completed during initialization.

The remaining ~2.9s first-request overhead (vs 0.08s subsequent) is clearly not from auto-tuning of any kernels, since `TRITON_PRINT_AUTOTUNING=1` produces no output during the first request. The current suspicion is that this residual cost comes from Triton JIT compilation (as opposed to auto-tuning), but this requires further investigation as a follow-up.

<details>
<summary>Benchmark script</summary>

```python
#!/usr/bin/env python3
"""Measure first-request vs subsequent-request latency for a Mamba2 model."""
import os
import shutil
import time

for cache_dir in [
    os.path.expanduser("~/.triton/cache"),
    os.path.expanduser("~/.cache/vllm/torch_compile_cache"),
]:
    if os.path.exists(cache_dir):
        shutil.rmtree(cache_dir)
        print(f"Cleared {cache_dir}")

from vllm import LLM, SamplingParams

NUM_REQUESTS = 5
PROMPT = "The capital of France is"

print("Loading model...")
t0 = time.perf_counter()
llm = LLM(model="nvidia/NVIDIA-Nemotron-3-Nano-4B-BF16",
           max_model_len=512, trust_remote_code=True)
load_time = time.perf_counter() - t0
print(f"Model loaded in {load_time:.1f}s\n")

sampling_params = SamplingParams(temperature=0.0, max_tokens=20)

results = []
for i in range(NUM_REQUESTS):
    t_start = time.perf_counter()
    llm.generate([PROMPT], sampling_params)
    latency = time.perf_counter() - t_start
    results.append(latency)
    print(f"  Request {i+1}: {latency:.3f}s")

first = results[0]
rest_avg = sum(results[1:]) / len(results[1:])
print(f"\n  Model load time:    {load_time:.1f}s")
print(f"  First request:      {first:.3f}s")
print(f"  Subsequent avg:     {rest_avg:.3f}s")
print(f"  Ratio (1st / avg):  {first / rest_avg:.2f}x")
```

</details>

## Test plan

- [x] Verified autotuning moves from first request to model init via `TRITON_PRINT_AUTOTUNING=1`
- [x] Confirmed first-request latency drops from 31.3s to 2.9s
- [x] Confirmed subsequent request latency is unchanged
- [ ] Run existing Mamba2 model tests

### CODE REVIEW

File: vllm/model_executor/layers/mamba/mamba_mixer2.py [hunk 1]
```diff
@@ -556,6 +559,106 @@ def forward(
...
+                    "Mamba2 SSD kernel warmup failed for layer %s "
+                    "(initial_states=%s). First inference may experience "
+                    "latency spike or OOM due to autotuner.",
+                    self.prefix,
+                    use_initial_states,
+                    exc_info=True,
+                )
+
+        logger.info("Mamba2 SSD kernel warmup completed for layer %s", self.prefix)
+        torch.accelerator.empty_cache()
```
[the remark this answers is not in the document]
Comment (tomeras91) [reply vllm/model_executor/layers/mamba/mamba_mixer2.py:661]: wrong recommendation.
this is the standard idiom in vLLM

File: vllm/model_executor/layers/mamba/mamba_mixer2.py [hunk 2]
```diff
@@ -556,6 +559,106 @@ def forward(
...
+
+        # Triton's autotuner includes tensor dtypes in its cache key,
+        # so state_dtype must match what real inference uses.
+        _, ssm_state_dtype = self.get_state_dtype()
+
+        seqlen = chunk_size
+        batch = 1
+        nchunks = seqlen // chunk_size  # = 1
+
+        x = torch.zeros(seqlen, nheads, headdim, device=device, dtype=dtype)
```
Comment (tomeras91) [inline vllm/model_executor/layers/mamba/mamba_mixer2.py]: I think I prefer we use randn here instead of zeros, to avoid any chance of invocating any fast-path for zeros in the kernels..

File: vllm/model_executor/layers/mamba/mamba_mixer2.py [hunk 3]
```diff
@@ -556,6 +559,106 @@ def forward(
...
+        headdim = self.head_dim
+        dstate = self.ssm_state_size
+
+        chunk_size = (
+            self.model_config.get_mamba_chunk_size()
+            if self.model_config is not None
+            else None
+        )
+        if chunk_size is None:
+            chunk_size = 64
```
Comment (tomeras91) [inline vllm/model_executor/layers/mamba/mamba_mixer2.py]: I think that if `self.model_config` is not defined, we should skip autotuning because:
1. This should never happen when the mamba block is initialized as part of a model
2. chunk_size has to be defined during actual forward pass, so if there's a mismatch between the chunk_size used in warmup autotuning and the one used during actual forward pass, warmup autotuning becomes irrelevant, and autotuning will be done again during actual forward pass.

In any case, 64 is not a good default for chunk size. The actual default in `ModelConfig.get_mamba_chunk_size` is 2048: 
https://github.com/vllm-project/vllm/blob/b39c266dae8cd7aee31f667c973e9698ed0b2361/vllm/config/model.py#L1339
(BTW, the docstring there says "we use a default chunk size of 1024" but the code returns 2048.. maybe fix while we're at it)

File: vllm/model_executor/layers/mamba/mamba_mixer2.py [hunk 4]
```diff
@@ -556,6 +559,106 @@ def forward(
 
         return output
 
+    def _warmup_ssd_kernels(self, projected_states: torch.Tensor) -> None:
+        """Run a minimal SSD forward pass to trigger Triton autotuning
+        while GPU memory is still plentiful (before SSM cache allocation).
+        """
+        if hasattr(self, "_ssd_kernels_warmed_up"):
```
Comment (tomeras91) [inline vllm/model_executor/layers/mamba/mamba_mixer2.py]: nit: Doesn't it make more sense to initialize this attribute in `__init__` as false and here just have `if not self._ssd_kernels_warmed_up`?
I realize what you have here is the same pattern #36599 used for GDN.. but it's still wierd. Maybe we should change it there as well.

File: vllm/model_executor/layers/mamba/mamba_mixer2.py [hunk 5]
```diff
@@ -556,6 +559,106 @@ def forward(
...
+            return
+        self._ssd_kernels_warmed_up = True
+        logger.info("Starting Mamba2 SSD autotuning warmup for layer %s", self.prefix)
```
Comment (tomeras91) [inline vllm/model_executor/layers/mamba/mamba_mixer2.py]: nit: This might clutter the logs as it is printed for every mamba block in the model.. maybe we can move to `logger.debug` and have just one `logger.info` call for the entire model?
Also makes sense since just the first layer will actually run auto-tuning and next layers will just hit the cache as the autotuning keys don't change between layers (assuming mamba dims are constant across the model..)

File: vllm/model_executor/layers/mamba/mamba_mixer2.py [hunk 6]
```diff
@@ -556,6 +559,106 @@ def forward(
...
+            device=device,
+            dtype=torch.int32,
+        )
+        last_chunk_indices = torch.tensor(
+            [nchunks - 1], device=device, dtype=torch.int32
+        )
+        seq_idx = torch.zeros(nchunks, device=device, dtype=torch.int32)
+        out = torch.empty(seqlen, nheads, headdim, device=device, dtype=dtype)
+
+        # Two kernels (_state_passing_fwd, _chunk_scan_fwd) use
```
Comment (tomeras91) [inline vllm/model_executor/layers/mamba/mamba_mixer2.py:612]: nit: this comment is slightly misleading. Technically, `HAS_INITSTATES` is not in the autotune key, so it's not that a different config is needed based on the value of `HAS_INITSTATES`. The reason this loop is needed is just to JIT-compile the kernel for both cases

**Review by tomeras91 — COMMENTED** [review]
Thanks @tdoublep for taking this! This is very helpful!

Added a few comments

A general comment -  Can we add a sentence somewhere saying SSD kernels don't have seqlen/batch-size dependent autotune keys? To preempt the obvious question "why didn't you autotune for different seqlens and batch sizes"..

File: vllm/model_executor/layers/mamba/mamba_mixer2.py (hunk 6 above)
Comment (tdoublep) [reply vllm/model_executor/layers/mamba/mamba_mixer2.py:612]: Correct - done

File: vllm/model_executor/layers/mamba/mamba_mixer2.py (hunk 2 above)
Comment (tdoublep) [reply vllm/model_executor/layers/mamba/mamba_mixer2.py]: Makes sense, done.

File: vllm/model_executor/layers/mamba/mamba_mixer2.py (hunk 3 above)
Comment (tdoublep) [reply vllm/model_executor/layers/mamba/mamba_mixer2.py]: I had it return if the model config isn't defined, and changed the comment in the config.

File: vllm/model_executor/layers/mamba/mamba_mixer2.py (hunk 5 above)
Comment (tdoublep) [reply vllm/model_executor/layers/mamba/mamba_mixer2.py]: Changed it to use info_once (and debug for each layer)

File: vllm/model_executor/layers/mamba/mamba_mixer2.py (hunk 4 above)
Comment (tdoublep) [reply vllm/model_executor/layers/mamba/mamba_mixer2.py]: I agree; Changed it for GDN too .

**Comment by tdoublep** [comment]
@tomeras91 Thanks for the review - I think I have addressed all feedback, please TAL.

**Review by tomeras91 — APPROVED** [review]
Thanks @tdoublep 
LGTM!
